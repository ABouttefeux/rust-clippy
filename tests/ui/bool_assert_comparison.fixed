// run-rustfix

#![warn(clippy::bool_assert_comparison)]

use std::ops::Not;

macro_rules! a {
    () => {
        true
    };
}
macro_rules! b {
    () => {
        true
    };
}

// Implements the Not trait but with an output type
// that's not bool. Should not suggest a rewrite
#[derive(Debug, Copy, Clone)]
#[allow(dead_code)]
enum ImplNotTraitWithoutBool {
    VariantX(bool),
    VariantY(u32),
}

impl PartialEq<bool> for ImplNotTraitWithoutBool {
    fn eq(&self, other: &bool) -> bool {
        match *self {
            ImplNotTraitWithoutBool::VariantX(b) => b == *other,
            _ => false,
        }
    }
}

impl Not for ImplNotTraitWithoutBool {
    type Output = Self;

    fn not(self) -> Self::Output {
        match self {
            ImplNotTraitWithoutBool::VariantX(b) => ImplNotTraitWithoutBool::VariantX(!b),
            ImplNotTraitWithoutBool::VariantY(0) => ImplNotTraitWithoutBool::VariantY(1),
            ImplNotTraitWithoutBool::VariantY(_) => ImplNotTraitWithoutBool::VariantY(0),
        }
    }
}

// This type implements the Not trait with an Output of
// type bool. Using assert!(..) must be suggested
#[derive(Debug, Clone, Copy)]
struct ImplNotTraitWithBool;

impl PartialEq<bool> for ImplNotTraitWithBool {
    fn eq(&self, _other: &bool) -> bool {
        false
    }
}

impl Not for ImplNotTraitWithBool {
    type Output = bool;

    fn not(self) -> Self::Output {
        true
    }
}

fn main() {
    let a = ImplNotTraitWithoutBool::VariantX(true);
    let b = ImplNotTraitWithBool;

    assert_eq!("a".len(), 1);
    assert!(!("a".is_empty()));
    assert!("".is_empty());
    assert!("".is_empty());
    assert_eq!(a!(), b!());
    assert_eq!(a!(), "".is_empty());
    assert_eq!("".is_empty(), b!());
    assert_eq!(a, true);
    assert!(b);

    assert_ne!("a".len(), 1);
    assert!("a".is_empty());
    assert!(!("".is_empty()));
    assert!(!("".is_empty()));
    assert_ne!(a!(), b!());
    assert_ne!(a!(), "".is_empty());
    assert_ne!("".is_empty(), b!());
    assert_ne!(a, true);
    assert!(!(b));

    debug_assert_eq!("a".len(), 1);
    debug_assert!(!("a".is_empty()));
    debug_assert!("".is_empty());
    debug_assert!("".is_empty());
    debug_assert_eq!(a!(), b!());
    debug_assert_eq!(a!(), "".is_empty());
    debug_assert_eq!("".is_empty(), b!());
    debug_assert_eq!(a, true);
    debug_assert!(b);

    debug_assert_ne!("a".len(), 1);
    debug_assert!("a".is_empty());
    debug_assert!(!("".is_empty()));
    debug_assert!(!("".is_empty()));
    debug_assert_ne!(a!(), b!());
    debug_assert_ne!(a!(), "".is_empty());
    debug_assert_ne!("".is_empty(), b!());
    debug_assert_ne!(a, true);
    debug_assert!(!(b));

    // assert with error messages
    assert_eq!("a".len(), 1, "tadam {}", 1);
    assert_eq!("a".len(), 1, "tadam {}", true);
    assert!(!("a".is_empty()), "tadam {}", 1);
    assert!(!("a".is_empty()), "tadam {}", true);
    assert!(!("a".is_empty()), "tadam {}", true);
    assert_eq!(a, true, "tadam {}", false);
    assert!("a".is_empty(), "tadam {} {}", false, 6);

    debug_assert_eq!("a".len(), 1, "tadam {}", 1);
    debug_assert_eq!("a".len(), 1, "tadam {}", true);
    debug_assert!(!("a".is_empty()), "tadam {}", 1);
    debug_assert!(!("a".is_empty()), "tadam {}", true);
    debug_assert!(!("a".is_empty()), "tadam {}", true);
    debug_assert_eq!(a, true, "tadam {}", false);
    debug_assert!("a".is_empty(), "tadam {} {}", false, "b");

    // Without ; at the end
    {
        debug_assert!(!("a".is_empty()), "tadam {}", true)
    };
    {
        assert_eq!("a".len(), 1, "tadam {}", 1)
    };
    {
        assert_ne!("a".len(), 1, "tadam {}", true)
    };
}
