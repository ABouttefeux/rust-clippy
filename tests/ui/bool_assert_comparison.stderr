error: used `assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:72:5
   |
LL |     assert_eq!("a".is_empty(), false);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!(!("a".is_empty()))`
   |
   = note: `-D clippy::bool-assert-comparison` implied by `-D warnings`

error: used `assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:73:5
   |
LL |     assert_eq!("".is_empty(), true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!("".is_empty())`

error: used `assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:74:5
   |
LL |     assert_eq!(true, "".is_empty());
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!("".is_empty())`

error: used `assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:79:5
   |
LL |     assert_eq!(b, true);
   |     ^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!(b)`

error: used `assert_ne!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:82:5
   |
LL |     assert_ne!("a".is_empty(), false);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!("a".is_empty())`

error: used `assert_ne!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:83:5
   |
LL |     assert_ne!("".is_empty(), true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!(!("".is_empty()))`

error: used `assert_ne!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:84:5
   |
LL |     assert_ne!(true, "".is_empty());
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!(!("".is_empty()))`

error: used `assert_ne!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:89:5
   |
LL |     assert_ne!(b, true);
   |     ^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!(!(b))`

error: used `debug_assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:92:5
   |
LL |     debug_assert_eq!("a".is_empty(), false);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!(!("a".is_empty()))`

error: used `debug_assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:93:5
   |
LL |     debug_assert_eq!("".is_empty(), true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!("".is_empty())`

error: used `debug_assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:94:5
   |
LL |     debug_assert_eq!(true, "".is_empty());
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!("".is_empty())`

error: used `debug_assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:99:5
   |
LL |     debug_assert_eq!(b, true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!(b)`

error: used `debug_assert_ne!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:102:5
   |
LL |     debug_assert_ne!("a".is_empty(), false);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!("a".is_empty())`

error: used `debug_assert_ne!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:103:5
   |
LL |     debug_assert_ne!("".is_empty(), true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!(!("".is_empty()))`

error: used `debug_assert_ne!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:104:5
   |
LL |     debug_assert_ne!(true, "".is_empty());
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!(!("".is_empty()))`

error: used `debug_assert_ne!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:109:5
   |
LL |     debug_assert_ne!(b, true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!(!(b))`

error: used `assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:114:5
   |
LL |     assert_eq!("a".is_empty(), false, "tadam {}", 1);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!(!("a".is_empty()), "tadam {}", 1)`

error: used `assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:115:5
   |
LL |     assert_eq!("a".is_empty(), false, "tadam {}", true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!(!("a".is_empty()), "tadam {}", true)`

error: used `assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:116:5
   |
LL |     assert_eq!(false, "a".is_empty(), "tadam {}", true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!(!("a".is_empty()), "tadam {}", true)`

error: used `assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:118:5
   |
LL |     assert_eq!("a".is_empty(), true, "tadam {} {}", false, 6);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `assert!("a".is_empty(), "tadam {} {}", false, 6)`

error: used `debug_assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:122:5
   |
LL |     debug_assert_eq!("a".is_empty(), false, "tadam {}", 1);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!(!("a".is_empty()), "tadam {}", 1)`

error: used `debug_assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:123:5
   |
LL |     debug_assert_eq!("a".is_empty(), false, "tadam {}", true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!(!("a".is_empty()), "tadam {}", true)`

error: used `debug_assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:124:5
   |
LL |     debug_assert_eq!(false, "a".is_empty(), "tadam {}", true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!(!("a".is_empty()), "tadam {}", true)`

error: used `debug_assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:126:5
   |
LL |     debug_assert_eq!("a".is_empty(), true, "tadam {} {}", false, "b");
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!("a".is_empty(), "tadam {} {}", false, "b")`

error: used `debug_assert_eq!` with a literal bool
  --> $DIR/bool_assert_comparison.rs:130:9
   |
LL |         debug_assert_eq!(false, "a".is_empty(), "tadam {}", true)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert!(!("a".is_empty()), "tadam {}", true)`

error: aborting due to 25 previous errors

